src/main/java/org/roaringbitmap/longlong/Roaring64NavigableMap.java

```diff
@@ -0,0 +1,1351 @@
+        // is excluded
+        endLowAsLong = Util.toUnsignedLong(-1) + 1;
+      }
+
+      if (endLowAsLong > startLowAsLong) {
+        // Initialize the bitmap only if there is access data to write
+        BitmapDataProvider bitmap = highToBitmap.get(high);
+        if (bitmap == null) {
+          bitmap = new MutableRoaringBitmap();
+          pushBitmapForHigh(high, bitmap);
+        }
+
+        if (bitmap instanceof RoaringBitmap) {
+          ((RoaringBitmap) bitmap).add(startLowAsLong, endLowAsLong);
+        } else if (bitmap instanceof MutableRoaringBitmap) {
+          ((MutableRoaringBitmap) bitmap).add(startLowAsLong, endLowAsLong);
+        } else {
+          throw new UnsupportedOperationException("TODO. Not for " + bitmap.getClass());
+        }
+      }
+    }
+
+    invalidateAboveHigh(startHigh);
+  }
+
+  @Override
+  public LongIterator getReverseLongIterator() {
+    return toIterator(highToBitmap.descendingMap().entrySet().iterator(), true);
+  }
+
+  @Override
+  public void add(long x) {
+    addLong(x);
+  }
+
+  @Override
+  public void remove(long x) {
+    int high = high(x);
+
+    BitmapDataProvider bitmap = highToBitmap.get(high);
+
+    if (bitmap != null) {
+      int low = low(x);
+      bitmap.remove(low);
+
+      // Invalidate only if actually modified
+      invalidateAboveHigh(high);
+    }
+
+  }
+
+  @Override
+  public void trim() {
+    for (BitmapDataProvider bitmap : highToBitmap.values()) {
+      bitmap.trim();
+    }
+  }
+
+  @Override
+  public int hashCode() {
+    return highToBitmap.hashCode();
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (this == obj) {
+      return true;
+    }
+    if (obj == null) {
+      return false;
+    }
+    if (getClass() != obj.getClass()) {
+      return false;
+    }
+    Roaring64NavigableMap other = (Roaring64NavigableMap) obj;
+    return Objects.equals(highToBitmap, other.highToBitmap);
+  }
+
+
+
+  /**
+   * Add the value if it is not already present, otherwise remove it.
+   *
+   * @param x long value
+   */
+  public void flip(final long x) {
+    int high = RoaringIntPacking.high(x);
+    BitmapDataProvider lowBitmap = highToBitmap.get(high);
+    if (lowBitmap == null) {
+      // The value is not added: add it without any flip specific code
+      addLong(x);
+    } else {
+      int low = RoaringIntPacking.low(x);
+
+      // .flip is not in BitmapDataProvider contract
+      // TODO Is it relevant to calling .flip with a cast?
+      if (lowBitmap instanceof RoaringBitmap) {
+        ((RoaringBitmap) lowBitmap).flip(low);
+      } else if (lowBitmap instanceof MutableRoaringBitmap) {
+        ((MutableRoaringBitmap) lowBitmap).flip(low);
+      } else {
+        // Fallback to a manual flip
+        if (lowBitmap.contains(low)) {
+          lowBitmap.remove(low);
+        } else {
+          lowBitmap.add(low);
+        }
+      }
```

Comment:
This branch of `#flip(...)` does not call `#invalidateAboveHigh(...)`, leading to invalid cardinalities and more.
See:

```java
  @Test
  public void testMe() {
    final Roaring64NavigableMap r = newUnsignedHeap();
    final long value = 1L;
    r.add(value);
    Assert.assertEquals(1, r.getLongCardinality());

    r.flip(1);
    Assert.assertEquals(0, r.getLongCardinality());
  }
```

Reply:
Also, I suggested earlier in the conversation that we could do a quick optimization when update the latest sortedHighs. 
I was thinking about something like this - idea applied to the manual part of flip.
```java
        // Fallback to a manual flip
        boolean contains = lowBitmap.contains(low);
        if (contains) {
          lowBitmap.remove(low);
        } else {
          lowBitmap.add(low);
        }

        if (allValid && sortedHighs.length > 0 && high == sortedHighs[sortedHighs.length - 1]) {
          // Directly update the cumulative count
          sortedCumulatedCardinality[high] += contains ? 1 : -1; 
        }
```
But I understand it requires that `add`, `remove` and `flip` return a boolean to indicate the results of the operations, which is not currently the case. I would understand if you want to keep that for another dev.

Reply:
I considered returning true for add, remove and flip. @lemire provided an alternative to .add which returns a boolean but it seemed slightly more expensive. I need to check if it exists for .remove and .flip. I decided to keep the discussion around this for a another PR.

About maintaining manually the cardinalities, we maintain them even if high is not the last high (by increasing all greater high). Needing more thinking and opinions, I keep thinking this should be reserved for another PR. Thanks for pushing it any way.

Reply:
That seems a very good choice. I was mentioning it not to miss anything in the dev. But as the needed methods are not there yet, it is better to wait for the next iteration on that component.